zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,339 @@
+		{
+			allocation = (struct active_allocation*)
+				CONTAINER_OF(iterator,struct active_allocation, node)->allocation_ptr;
 			break;
-		case MEM_REG_ATTR_SHARED_HEAP:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	if (allocation == NULL) return -EEXIST;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++)
+	{
+		if (is_ptr_in_range(ptr,
+			heap->physical_blocks_allocators[i]->buffer,
+			heap->physical_blocks_allocators[i]->num_blocks * (heap->physical_blocks_allocators[i]->blk_sz_shift^2)))
+		{
+			if (heap->span)
+			{
+				retval = sys_mem_blocks_free_contiguous(heap->physical_blocks_allocators[i], ptr, allocation->allocation_size);
+			}
+			else
+			{
+				retval = sys_mem_blocks_free(heap->physical_blocks_allocators[i], allocation->allocation_size, (void **)&ptr);
+			}
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	retval = sys_slist_find_and_remove(&heap->active_allocations, &allocation->node);
+
+	//SCAN MAPPINGS IF ANYTHING CHANGED and unmapping happend
+	sys_snode_t *iterator_phys;
+	sys_snode_t *iterator_activ;
+	SYS_SLIST_FOR_EACH_NODE(&physical_mappings, iterator_phys)
+	{
+		bool can_free = true;
+		SYS_SLIST_FOR_EACH_NODE(&heap->active_allocations, iterator_activ)
+		{
+			if (is_range_in_range(
+				(uint8_t *)CONTAINER_OF(iterator_activ,struct active_allocation, node)->allocation_ptr,
+				CONTAINER_OF(iterator_activ,struct active_allocation, node)->allocation_size,
+				(uint8_t *)CONTAINER_OF(iterator_phys,struct active_physical_mapping, node)->mapping_ptr,
+				CONTAINER_OF(iterator_phys,struct active_physical_mapping, node)->mapping_size))
+			{
+				can_free = false;
+				break;
+			}
+		}
+		if (can_free)
+		{
+			sys_mm_drv_unmap_region(CONTAINER_OF(iterator_phys,struct active_physical_mapping, node)->mapping_ptr, 
+				CONTAINER_OF(iterator_phys,struct active_physical_mapping, node)->mapping_size);
+			retval = sys_slist_find_and_remove(&physical_mappings, iterator_phys);
 			break;
-		default:
-			return -EINVAL;
 		}
 	}
 
+	return retval;
+}
+
+struct virtual_memory_heap *reconfigure_heap(struct virtual_memory_heap *heap, struct heap_config* cfg, bool span)
+{
+	struct virtual_memory_heap * retval = NULL;
+	uint32_t region_attribute = heap->virtual_region->attr;
+	if(free_heap(heap)) return retval;
+	return init_heap(cfg, region_attribute, span);
+}
+
+/**
+ * @brief Get default configuration for heap
+ *
+ * This will return config created based on region provided.
+ * It will try to split memory in even chunks and split it into
+ * bundles of size 64,128,...,1024
+ *
+ * @param region Virtual region which resources we will be using for heap
+ *
+ * @retval Returns default heap_config
+ */
+struct heap_config get_default_heap_config(struct sys_mm_drv_region *region)
+{
+	int buffer_size = 0.2 * region->size;
```

Comment:
Please stay with integer arithmetics. Why do you need a fifth of the size? Ah, I get it. It's for 64, 128, 256, 512 and 1024 byte allocations. Well, even if we decide to keep this, at least replace it with "/ 5"